docs/tutorials/hyperparameter_tuning/hp_tuning_with_optuna.md

```diff
@@ -0,0 +1,328 @@
+test_series.plot()
+plt.axvline(train_series.index[-1], color='r') # end of train dataset
+plt.grid(which="both")
+plt.legend(["test series", "end of train series"], loc="upper left")
+plt.show()
+```
+
+
+```python
+print(f"Length of forecasting window in test dataset: {len(test_series) - len(train_series)}")
+```
+
+## tuning parameters of DeepAR estimator
+
+
+```python
+import optuna
+import torch
+from gluonts.torch.model.deepar import DeepAREstimator
+from gluonts.mx import Trainer
+from gluonts.evaluation import Evaluator
+```
+
+An example that tune the DeepAR estimator on elctricity dataset via Optuna. We choose two hyperparameters num_layers and hidden_size to optimize.
+
+Define an Objective class used in tuning process of Optuna.
+With `__init__` function, initialize the dataset, prediction_length, freq and metric_type(means what metric to be used to evaluate the DeepAREstimator).
+With `get_params` function, define what hyperparameters to be tuned within given range.
+With `split_entry` function, split each time series of the dataset into two part:
+- entry_past: the training part 
+- entry_future: the label part used in validation
+With `__call__` function, define the way DeepAREstimator is used in training and validation.
+
+
+```python
+class Objective:  
+    def __init__(self, dataset, prediction_length, freq, metric_type="mean_wQuantileLoss"):
+        self.dataset = dataset
+        self.prediction_length = prediction_length
+        self.freq = freq
+        self.metric_type = metric_type
+    
+    def get_params(self, trial) -> dict:
+        return {
+            "num_layers": trial.suggest_int("num_layers", 1, 5),
+            "hidden_size": trial.suggest_int("hidden_size", 10, 50),
+        }
+
+    def split_entry(self, entry):
+        entry_past = {}
+        for key, value in entry.items():
+            if key == "target":
+                entry_past[key] = value[: len(value) - self.prediction_length]
+            else:
+                entry_past[key] = value
+
+        df = pd.DataFrame(entry['target'], columns=[entry['item_id']])
+        df = df.set_index(pd.date_range(start=entry['start'], 
+                                        periods=len(entry['target']), freq=self.freq))
+        return entry_past, df[-self.prediction_length:]
+     
+    def __call__(self, trial):
+        params = self.get_params(trial)
+        estimator = DeepAREstimator(
+            num_layers=params['num_layers'],
+            hidden_size=params['hidden_size'],
+            prediction_length=self.prediction_length, 
+            freq=self.freq,
+            trainer_kwargs={
+                "progress_bar_refresh_rate": 0, 
+                "weights_summary": None, 
+                "max_epochs": 5, 
+            }
+        )
+        
+        entry_splited = [self.split_entry(entry) for entry in self.dataset]
+        entry_pasts = [entry[0] for entry in entry_splited]
+        entry_futures = [entry[1] for entry in entry_splited]
```

Comment:
Maybe this split could be done just once when the objective is constructed? That is, when `__init__` is called, instead of in `__call__`

Reply:
OK,I couldn't agree with you any more and I will modify this.